lib/icinga/dependency.cpp

```diff
@@ -9,106 +9,165 @@
 #include "base/exception.hpp"
 #include <map>
 #include <sstream>
+#include <unordered_map>
 #include <utility>
+#include <variant>
 
 using namespace icinga;
 
 REGISTER_TYPE(Dependency);
 
-bool Dependency::m_AssertNoCyclesForIndividualDeps = false;
+INITIALIZE_ONCE(&Dependency::StaticInitialize);
 
-struct DependencyCycleNode
+void Dependency::StaticInitialize()
 {
-	bool Visited = false;
-	bool OnStack = false;
-};
-
-struct DependencyStackFrame
-{
-	ConfigObject::Ptr Node;
-	bool Implicit;
-
-	inline DependencyStackFrame(ConfigObject::Ptr node, bool implicit = false) : Node(std::move(node)), Implicit(implicit)
-	{ }
-};
-
-struct DependencyCycleGraph
-{
-	std::map<Checkable::Ptr, DependencyCycleNode> Nodes;
-	std::vector<DependencyStackFrame> Stack;
-};
-
-static void AssertNoDependencyCycle(const Checkable::Ptr& checkable, DependencyCycleGraph& graph, bool implicit = false);
+	dynamic_cast<ConfigType*>(Dependency::TypeInstance.get())->BeforeOnAllConfigLoaded.connect(&BeforeOnAllConfigLoadedHandler);
```

Comment:
There is also `ConfigType::Get<Dependency>()`, which performs an equivalent type casting, I think it can be used here.

Reply:
Oh, that looks so much nicer :+1:

Reply:
Don't ask me how. I know I had this change locally but it got lost somewhere, probably when reverting some other changes I tried with that new `ConfigItems` class.